### PR TITLE
Task 967

### LINKED ISSUE #967: Explore DVFS inefficiencies and impact on instruction per cycle rate
> estimate 5

The Energy Efficient agent tests (test_ee_short_region_slop) have shown some impact to IPC when DVFS is used.  

Done criterion: Write scripts for producing reference data for future performance tests that explore this effect and aid in understanding its root cause.
- Using two-region app from #947 and measure IPC in trace; produce plot(s) of the following metrics over time: IPC, requested frequency, achieved core frequency, uncore frequency, MSR::PERF_CTL:FREQ  
- Rerun existing script with ioctl writes and produce same set of plot from #944 for comparison.

Parent story #978

### PR DESCRIPTION
**Pull request by cmcantalupo**
- Fixes #967 from github issues
- Adds trace generation and plotting for test_ee_short_region_slop test

### CODE REVIEW

File: test_integration/test_ee_short_region_slop.cpp [hunk 1]
```diff
@@ -50,7 +50,7 @@ int main(int argc, char **argv)
     bool is_verbose = false;
     geopm::Profile &prof = geopm::Profile::default_profile();
     // Run twelve trials with region duration ranging from 100 us - 400 ms
```
Comment (dianarg) [inline test_integration/test_ee_short_region_slop.cpp]: comment is no longer correct

File: test_integration/test_ee_short_region_slop.py [hunk 2]
```diff
@@ -70,6 +72,13 @@
     from test_integration import util
     geopmpy.error.exc_clear()
 
+# Globals controlling uniform y axis limits in all plots
```
Comment (dianarg) [inline test_integration/test_ee_short_region_slop.py:82]: Could be moved into test class?  Plotting functions also would need to move

File: test_integration/test_ee_short_region_slop.py [hunk 3]
```diff
@@ -204,71 +241,89 @@ def tearDownClass(cls):
...
+                if len(in_path) == 0:
+                    raise RuntimeError('No trace files found')
+                in_path = in_path[0]
+                trace = read_trace(in_path)
+                # Get the first and last time that the regions of interest were executed
+                begin_time = trace['TIME'].loc[(trace['REGION_HASH'] == begin_region_hash) &
+                                               (trace['REGION_COUNT'] == region_count[0])].iloc[0]
+                end_time = trace['TIME'].loc[(trace['REGION_HASH'] == end_region_hash) &
+                                             (trace['REGION_COUNT'] == region_count[-1])].iloc[-1]
+                duration = end_time - begin_time
```
Comment (dianarg) [inline test_integration/test_ee_short_region_slop.py]: duration is a confusing name
just used for width of plot

File: test_integration/test_ee_short_region_slop.py [hunk 4]
```diff
@@ -325,11 +364,135 @@ def extract_data_region(report, cols, key):
...
+
+def select_time_range(trace, duration, begin_time=None):
+    if begin_time is None:
+        begin_time = trace['TIME'].iloc[0]
+    end_time = begin_time + duration
+    return trace.loc[(trace['TIME'] >= begin_time) &
+                     (trace['TIME'] < end_time)]
+
+
+def generate_trace_region_plot(trace, out_path, region_name, region_count):
```
Comment (dianarg) [inline test_integration/test_ee_short_region_slop.py]: add "overlay" to name?

File: test_integration/test_scaling_region.py [hunk 5]
```diff
@@ -93,14 +94,8 @@ def setUpClass(cls):
...
         # scaling_region_0, scaling_region_1, ... , scaling_region_30
-        cls._region_hash = [0x1f2b4ef8, 0xed40cdfb, 0xfe103e0f, 0x0c7bbd0c,
-                            0xd8b1d9e7, 0x2ada5ae4, 0x398aa910, 0xcbe12a13,
-                            0x95f21637, 0x67999534, 0xea0548e2, 0xa33935c5,
-                            0x787db2ac, 0x3141cf8b, 0xcb18ca8f, 0x8224b7a8,
-                            0x596030c1, 0x105c4de6, 0xa83e4c38, 0xe102311f,
-                            0xf3aa44cb, 0xba9639ec, 0x61d2be85, 0x28eec3a2,
-                            0xd2b7c6a6, 0x9b8bbb81, 0x40cf3ce8, 0x09f341cf,
-                            0xb1914011, 0xf8ad3d36, 0x07949283]
+        cls._region_hash = [geopmpy.hash.crc32_str('scaling_region_{}'.format(idx))
```
Comment (dianarg) [inline test_integration/test_scaling_region.py:97]: yay

File: test_integration/test_ee_short_region_slop.py [hunk 6]
```diff
@@ -325,11 +364,135 @@ def extract_data_region(report, cols, key):
...
+def read_trace(path):
+    converters={'REGION_HASH': lambda xx: int(xx, 16)}
+    return pandas.read_csv(path, sep='|', comment='#', converters=converters)
+
+
+def get_ipc(trace):
+    di = trace['INSTRUCTIONS_RETIRED'].diff()
+    dc = trace['CYCLES_THREAD'].diff()
+    return di / dc
+
```
Comment (bgeltz) [inline test_integration/test_ee_short_region_slop.py:432]: need another \n

File: test_integration/test_ee_short_region_slop.py [hunk 7]
```diff
@@ -325,11 +364,135 @@ def extract_data_region(report, cols, key):
...
+def get_freq(trace):
+    return trace['FREQUENCY']
+
```
Comment (bgeltz) [inline test_integration/test_ee_short_region_slop.py:436]: need another \n

File: test_integration/test_ee_short_region_slop.py [hunk 8]
```diff
@@ -325,11 +364,135 @@ def extract_data_region(report, cols, key):
...
+def select_region_count(trace, region_count):
+    return trace.loc[trace['REGION_COUNT'] == region_count]
+
+
+def select_region(trace, region_prefix, extension=''):
+    """e.g.:
+    >>> select_region(trace, 'scaling_2')
+    >>> select_region(trace, 'scaling_', range(12))
+    >>> select_region(trace, '', ['scaling_3', 'timed_3'])
```
Comment (bgeltz) [inline test_integration/test_ee_short_region_slop.py:458]: Nice.

File: test_integration/test_ee_short_region_slop.py [hunk 9]
```diff
@@ -325,11 +364,135 @@ def extract_data_region(report, cols, key):
...
+    if type(extension) is str:
+        extension = [extension]
+    region_hash = []
+    for ext in extension:
+        region_name = '{}{}'.format(region_prefix, ext)
+        region_hash.append(geopmpy.hash.crc32_str(region_name))
+    return trace.loc[trace['REGION_HASH'].isin(region_hash)]
+
+
+def select_time_range(trace, duration, begin_time=None):
```
Comment (bgeltz) [inline test_integration/test_ee_short_region_slop.py]: I like these helper funcs.  I'm wondering how we should organize a #import with all of them.

File: test_integration/test_ee_short_region_slop.py [hunk 10]
```diff
@@ -325,11 +364,135 @@ def extract_data_region(report, cols, key):
...
+        plt.xlabel('time since region start (sec)')
+        plt.ylabel('CPU freq (Hz)')
+        plt.ylim(g_plot_freq_lim)
+        freq_legend.append('count {}'.format(rc))
+    plt.subplot(2,1,1)
+    plt.legend(ipc_legend)
+    plt.subplot(2,1,2)
+    plt.legend(freq_legend)
+    plt.savefig(out_path)
+    plt.close()
```
Comment (bgeltz) [inline test_integration/test_ee_short_region_slop.py]: What do you have against \n's ? 😆

**Review by bgeltz — CHANGES_REQUESTED** [review]
This is pretty slick with all the plotting stuff.  Tiny whitespace changes needed, and could you add to the scaling model region patch commit msg some more about why you're doing that?

File: test_integration/test_ee_short_region_slop.py (hunk 2 above)
Comment (cmcantalupo) [reply test_integration/test_ee_short_region_slop.py:82]: I'm going to leave these separate for now with the idea that some other test may want to reuse the plotting routines.  We can make these bounds a default if that happens.

File: test_integration/test_ee_short_region_slop.py (hunk 9 above)
Comment (cmcantalupo) [reply test_integration/test_ee_short_region_slop.py]: Seems like a good idea for another PR.

File: test_integration/test_ee_short_region_slop.py (hunk 10 above)
Comment (cmcantalupo) [reply test_integration/test_ee_short_region_slop.py]: I have removed the patch that has the memory alignment change.  I tried to refactor the blob into a loop, but it wasn't worth the hassle.

**Review by dianarg — DISMISSED** [review]
I don't see any reason not to merge this as is now.  In the future, we might want to convert the sbatch script to a more generic launch script similar to the tutorials.

**Review by bgeltz — DISMISSED** [review]

**Review by dianarg — APPROVED** [review]

**Pull request merged**